numba/targets/arraymath.py

```diff
@@ -1352,6 +1352,28 @@ def _prepare_array_impl(arr):
     else:
         return lambda arr: _asarray(arr).ravel()
 
+def _dtype_of_compound(inobj):
+    obj = inobj
+    while True:
+        try:
+            if isinstance(obj, (types.Number, types.Boolean)):
+                return as_dtype(obj)
+            l = getattr(obj, '__len__', None)
+            if l is not None and l() == 0: # empty tuple or similar
+                return np.float64
+            dt = getattr(obj, 'dtype', None)
+            if dt is None:
+                raise TypeError("type has no dtype attr")
+            return as_dtype(dt)
+        except (TypeError, NotImplementedError) as e:
+            # has dtype, but is not something understandable
+            if 'cannot be represented' in str(e) or \
+            'data type not understood' in str(e):
+                obj = obj.dtype
```

Comment:
I think the only sensible type for `obj` is `Sequence`.  Better to explicitly check for that then to rely on exception messages.